README.rst
What is SDoN PoC
================
This repository contains a Proof of Concept (PoC) for our novel method of managing Overlay Networks by leveraging the SDN paradigm. As the architecture is explained in the accompanying research paper, the following steps are here to re-create our testbed.
This repository should be cloned in the host machine. 

Machines
========
We recommend using Linux virtual machines for this purpose. We used XEN as hypervisor and for convenience we have placed our XEN config file in sdon-poc/ryu/app/sdonManager/sdon1.cfg. This config file should be cloned for each of the other SDoN nodes, i.e. sdon node 2 to 6.

Installing and configuring XEN is not the scope of this README, however we have used the information from the following link to configure our XEN VMs: https://help.ubuntu.com/community/Xen , especially focusing on the section about "Manually Create a PV Guest VM".

As IP subnet, we chose 10.0.0.0/24, where 10.0.0.1 is the XEN bridge for the VMs, and the SDON nodes are at 10.0.0.101 to 10.0.0.106.

After each VM is up and running, one has to install Open vSwitch (ovs) and scapy in each VM. Then copy the python script at sdon-poc/app/sdonManager/sdonManager.py from this repository in the VM's home directory. Note this script is the Signaling Module in our paper.


Running the Ryu SDoN app
========================
The SDoN Controller application is at sdon-poc/app/sdon.py and should be run with:
	% ryu-manager ryu/app/sdon.py
from the folder where this code repository is cloned.

After running this command, the SDoN Controller waits for the SDoN nodes (i.e. your VMs) to connect and start their overlay application.

Go now to the console of the VMs and run on each VM the following command to start the signaling module:
	% sudo python sdonManager.py
and follow the on screen menu that will deliver a our PoC demo.



The remainder is from the original ryu controller README file.

What's Ryu
==========
Ryu is a component-based software defined networking framework.

Ryu provides software components with well defined API that make it
easy for developers to create new network management and control
applications. Ryu supports various protocols for managing network
devices, such as OpenFlow, Netconf, OF-config, etc. About OpenFlow,
Ryu supports fully 1.0, 1.2, 1.3, 1.4, 1.5 and Nicira Extensions.

All of the code is freely available under the Apache 2.0 license. Ryu
is fully written in Python.


Quick Start
===========
Installing Ryu is quite easy::

   % pip install ryu

If you prefer to install Ryu from the source code::

   % git clone git://github.com/osrg/ryu.git
   % cd ryu; python ./setup.py install

If you want to write your Ryu application, have a look at
`Writing ryu application <http://ryu.readthedocs.org/en/latest/writing_ryu_app.html>`_ document.
After writing your application, just type::

   % ryu-manager yourapp.py


Optional Requirements
=====================

Some functionalities of ryu requires extra packages:

- OF-Config requires lxml
- NETCONF requires paramiko
- BGP speaker (ssh console) requires paramiko

If you want to use the functionalities, please install requirements::

    % pip install lxml
    % pip install paramiko


Support
=======
Ryu Official site is `<http://osrg.github.io/ryu/>`_.

If you have any
questions, suggestions, and patches, the mailing list is available at
`ryu-devel ML
<https://lists.sourceforge.net/lists/listinfo/ryu-devel>`_.
`The ML archive at Gmane <http://dir.gmane.org/gmane.network.ryu.devel>`_
is also available.


ChangeLog
CHANGES
=======

* First commit. Based on original ryu 3.30


ryu/app/sdon.py
# Copyright (C) 2011 Nippon Telegraph and Telephone Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from ryu.base import app_manager
from ryu.controller import ofp_event
from ryu.controller.handler import CONFIG_DISPATCHER, MAIN_DISPATCHER
from ryu.controller.handler import set_ev_cls
from ryu.ofproto import ofproto_v1_3
from ryu.lib.packet import packet
from ryu.lib.packet import ethernet
from ryu.lib.packet import ether_types
from ryu.lib.packet import udp
from ryu.lib.packet import ipv4
from ryu.lib import addrconv
# from ryu.app import rest_router as rr
import struct

class Sdon(app_manager.RyuApp):
    OFP_VERSIONS = [ofproto_v1_3.OFP_VERSION]

    def __init__(self, *args, **kwargs):
        super(Sdon, self).__init__(*args, **kwargs)

        # keeps and ip to datapath, parser, OFport ID --  dictionary
        self.sdonID_to_OF = {}

        # A node's public IP address is also their SDoN node id
        self.sdonID_to_apps = {}

        # set static roles for the FT app
        self.sdonID_to_FTrole = {
            # S = srouce, F = forwarder, A = adder, D = destination
            "10.0.0.101": "[S]", # node A
            "10.0.0.102": "[S]", # node B
            "10.0.0.103": "[A]", # node C
            "10.0.0.104": "[D]", # node D
            "10.0.0.105": "[F]", # node E
            "10.0.0.106": "[D]"  # node F
        }

        self.sdonID_to_FTappIP = {
            "10.0.0.101": "192.168.101.1/24",
            "10.0.0.102": "192.168.101.2/24",
            "10.0.0.103": "192.168.101.3/24",
            "10.0.0.104": "192.168.101.4/24",
            "10.0.0.105": "192.168.101.5/24",
            "10.0.0.106": "192.168.101.6/24"
        }

        self.sdonID_to_tunnels = {
            "10.0.0.101": "[10.0.0.103;10.0.0.104]",
            "10.0.0.102": "[10.0.0.103;10.0.0.106]",
            "10.0.0.103": "[10.0.0.101;10.0.0.102;10.0.0.105]",
            "10.0.0.104": "[10.0.0.101;10.0.0.105]",
            "10.0.0.105": "[10.0.0.103;10.0.0.104;10.0.0.106]",
            "10.0.0.106": "[10.0.0.102;10.0.0.105]"
        }

        self.sdonID_to_getGo = {
            "10.0.0.101": None,
            "10.0.0.102": None,
            "10.0.0.103": None,
            "10.0.0.104": None,
            "10.0.0.105": None,
            "10.0.0.106": None
        }

        self.sdonID_to_rules = {
            "10.0.0.101": {2: [3,4]},

            "10.0.0.102": {2: [3,4]},

            "10.0.0.103": {2: [5],
                           3: [2],
                           4: [2]},

            "10.0.0.104": {3: [2],
                           4: [2]},

            "10.0.0.105": {3: [2],
                           2: [4,5]},

            "10.0.0.106": {3: [2],
                           4: [2]}
        }

        # This should be done in negotiation with the SDoN database in a server
        self.subnets_to_apps = {
            "FT": "192.168.101.0/24"
        }
    #
    @set_ev_cls(ofp_event.EventOFPPacketIn, MAIN_DISPATCHER)
    def _packet_in_handler(self, ev):
        print "PACKET IN!!!"
        # If you hit this you might want to increase
        # the "miss_send_length" of your switch
        if ev.msg.msg_len < ev.msg.total_len:
            self.logger.debug("packet truncated: only %s of %s bytes",
                              ev.msg.msg_len, ev.msg.total_len)
        msg = ev.msg
        datapath = msg.datapath
        print datapath
        ofproto = datapath.ofproto
        parser = datapath.ofproto_parser
        in_port = msg.match['in_port']
        print "In port: " + str(in_port)

        pkt = packet.Packet(msg.data)
        print pkt
        # self.logger.info("packet-in %s" % (pkt,))
        eth = pkt.get_protocols(ethernet.ethernet)[0]

        if eth.ethertype == ether_types.ETH_TYPE_LLDP:
            # ignore lldp packet
            return

        if eth.ethertype == ether_types.ETH_TYPE_IP:
            pkt_ip = pkt.get_protocol(ipv4.ipv4)
            if pkt_ip.dst == "1.2.3.4":
                print "Received SDoN signalling packet."
                data = str(pkt.protocols[-1])
                # sdonID = str(pkt_ip.src)
                print "Data: " + data
                self.handleSdonPacket(data, msg)


    def handleSdonPacket(self, data, msg):
        sdonID, action = data.split(",")
        sdonID = sdonID.strip()
        if action == "hello":
            self.handleSdonHello(sdonID, msg)
        if action == "getIp":
            self.handleGetIp(sdonID, msg)
        if action == "getRole":
            self.handleGetRole(sdonID, msg)
        if action == "getTunnels":
            self.handleGetTunnels(sdonID, msg)
        if action == "getRules":
            self.handleGetRules(sdonID, msg)
        if action == "getGo":
            self.handleGetGo(sdonID, msg)

    def handleSdonHello(self, sdonID, msg):
        print "SDoN node with pub IP: " + sdonID + " was added."
        self.sendReply(sdonID, msg, "hello,", "ack")
        print "Hello, ack sent back!"


    def handleGetIp(self, sdonID, msg):
        print "Got FT get IP request from:" + sdonID
        ip = self.sdonID_to_FTappIP[sdonID]
        self.sendReply(sdonID, msg, "setIp,", str(ip))

    def handleGetRole(self, sdonID, msg):
        print "Got FT get role request from:" + sdonID
        role = self.sdonID_to_FTrole[sdonID]
        self.sendReply(sdonID, msg, "setRoles,", str(role))

    def handleGetTunnels(self, sdonID, msg):
        print "Got getTunnels request from:" + sdonID
        tunnels = self.sdonID_to_tunnels[sdonID]
        self.sendReply(sdonID, msg, "setTunnels,", str(tunnels))

    def handleGetRules(self, sdonID, msg):
        print "Got getRules request from:" + sdonID
        self.sendReply(sdonID, msg, "setRules,", "ACK")
        self.setupRulesOnSdonNode(sdonID)

    def handleGetGo(self, sdonID, msg):
        print "Got getGo from: " + sdonID

        # This node is ready, set its flag to True
        self.sdonID_to_getGo[sdonID] = msg

        # Check if there is any other node left to ask to start,
        # if there is a node not chekced in, then don't give the go-ahead just yet
        for nId, origMgs in self.sdonID_to_getGo.items():
            if origMgs == None:
                return

        # if the function reached this point, it means the last node just checked in,
        # let the other know thay have to start trhei FT app
        for nId, origMgs in self.sdonID_to_getGo.items():
            self.sendReply(nId, origMgs, "setGo,", "ACK")

    def sendReply(self, sdonID, origMsg, action, reply):
        datapath = origMsg.datapath
        ofproto = datapath.ofproto
        parser = datapath.ofproto_parser
        in_port = origMsg.match['in_port']
        if (action == "hello,"):
            # save this SdonManager in the dictionary so we can contact that switch when we want
            print "Saving " + str(sdonID) + "'s OF details."
            self.sdonID_to_OF[sdonID] = [datapath, parser, in_port]

            # Initialize the dictionary of Apps available for an SdonNode
            # self.sdonID_to_apps[pubIp] = []

        # Reply to the action message
        pkt = packet.Packet()
        e = ethernet.ethernet()
        i = ipv4.ipv4(src="10.0.0.1", dst = "1.2.3.4")
        # u = udp.udp()
        pkt= e/i/(str(action) + str(reply))
        pkt.serialize()

        data = pkt.data
        actions=[parser.OFPActionOutput(in_port)]
        out = parser.OFPPacketOut(datapath=datapath,
                                  buffer_id=ofproto.OFP_NO_BUFFER,
                                  in_port=ofproto.OFPP_CONTROLLER,
                                  actions=actions,
                                  data=data)
        datapath.send_msg(out)

    def setupRulesOnSdonNode(self, sdonID):
        print self.sdonID_to_OF
        datapath = self.sdonID_to_OF[sdonID][0]
        ofproto = datapath.ofproto
        parser = datapath.ofproto_parser
        # Reply to the action message
        for inPort, outPorts in self.sdonID_to_rules[sdonID].items():
            print "In: ", inPort, "Outsputs: ", outPorts
            match = parser.OFPMatch(in_port=inPort)
            actions = []
            for outPort in outPorts:
                actions.append(parser.OFPActionOutput(outPort))

            self.add_flow(datapath, 1, match, actions)

        ## if out_port != ofproto.OFPP_FLOOD:
        ## match = parser.OFPMatch(ipv4_dst=ipv4_text_to_int(sdonID))
        # match = parser.OFPMatch(in_port=2) # this is the port of the FT app
        ## actions=[parser.OFPActionOutput(ofproto.OFPP_FLOOD)]
        # actions=[parser.OFPActionOutput(3), parser.OFPActionOutput(4)]
        # self.add_flow(datapath, 1, match, actions)


    def add_flow(self, datapath, priority, match, actions):
        ofproto = datapath.ofproto
        parser = datapath.ofproto_parser

        inst = [parser.OFPInstructionActions(ofproto.OFPIT_APPLY_ACTIONS,
                                             actions)]

        mod = parser.OFPFlowMod(datapath=datapath, priority=priority,
                                match=match, instructions=inst)
        datapath.send_msg(mod)


def ipv4_text_to_int(ip_text):
    if ip_text == 0:
        return ip_text
    assert isinstance(ip_text, str)
    return struct.unpack('!I', addrconv.ipv4.text_to_bin(ip_text))[0]


ryu/app/sdonManager/set_sdon_node.sh
#!/bin/bash

# Remove potentially existing "sdon" OVS switch
sudo ovs-vsctl del-br sdon
sudo ip link del veth1 type veth peer name veth2

# Add and "sdon" ovs switch
sudo ovs-vsctl add-br sdon

# Set "sdon" OF version to 1.3 so it can talk to Ryu
sudo ovs-vsctl set bridge sdon protocols=OpenFlow13

# Create an interface pair
sudo ip link add veth1 type veth peer name veth2

# Add veth1 to sdon
sudo ovs-vsctl add-port sdon veth1

# Set veth1 up
sudo ifconfig veth1 up

# Set ipv4 forwarding to Enabled
sudo sysctl -w net.ipv4.ip_forward=1

# Set ip for veth2
sudo ifconfig veth2 192.168.111.1/24

# Add tunnels to "sdon"
sudo ovs-vsctl add-port sdon gre1 -- set interface gre1 type=gre options:remote_ip=10.0.0.101 
sudo ovs-vsctl add-port sdon gre2 -- set interface gre2 type=gre options:remote_ip=10.0.0.102 

# Set controller for "sdon"
sudo ovs-vsctl set-controller sdon tcp:10.0.0.1:6633

# DONE!


ryu/app/sdonManager/sdonManager.py
#!/usr/bin/env python

import os
from scapy.all import *
import time
import random
import string
# import fileTransfer

CTRL_IP    = "10.0.0.1"
CTRL_PORT  = "6633"
CTRL_INTF  = "veth2"
APP_INTF   = "veth4"

# f = os.popen('ifconfig eth0 | grep "inet\ addr" | cut -d: -f2 | cut -d" " -f1')
MY_PUB_IP = ""

MY_IP = ""
MY_ROLE = ""
MY_ID = ""

print "Hello, this is the sdonManager app."

def mainMenu():
    while True:
        os.system("clear")
        print "=== Sdon Main menu ==="
        print MY_IP, MY_ID
        print "Options"
        print "--- Step by step ---"
        print "1. Create OVS switch and send Hello message to controller"
        print "2. Get  IP  my FT sdon application"
        print "3. Get role my FT sdon application"
        print "4. Setup my tunnels to neighbours"
        print "5. Setup my OF rules"
        print ""
        print "A. Do steps 1, 2, 3, 4, 5 sequentially"
        print "S. Start file transfer SDoN app"
        # print "9. Send test packet"
        print "0. Exit"
        choice = raw_input("Your choice: ")
        execOption(choice)

def execOption(choice):
    print "Your choice is: " ,choice
    if choice == "1":
        createOvsSwitch()
    elif choice == "2":
        getMyIp()
    elif choice == "3":
        getMyRole()
    elif choice == "4":
        getMyTunnels()
    elif choice == "5":
        getMyOfRules()
    elif choice == "9":
        sendTestPacket()
    # elif choice == "l":
    #     waitingForInstructions()
    elif choice == "A":
        createOvsSwitch()
        time.sleep(1.1)

        getMyIp()
        time.sleep(1.1)

        getMyRole()
        time.sleep(1.1)

        getMyTunnels()
        time.sleep(1.1)

        getMyOfRules()
        time.sleep(1.1)

    elif choice == "S":
        getGo()
    elif choice == "0":
        exit()
    raw_input("Press any key to continue...")
    return

def handleSdonSignalingPkt(pkt):
    packet = pkt[0]
    if packet == None:
        print "Timeout occured before server replied."
        return False
    ctrlReply = packet[IP].payload
    print ctrlReply
    handleControllerReply(str(ctrlReply))

    return True


def createOvsSwitch():
    print "Creating OVS switch."
    # Remove potentially existing "sdon" OVS switch
    os.system("sudo ovs-vsctl del-br sdon")
    os.system("sudo ovs-vsctl del-br mgm")
    os.system("sudo ip link del veth1 type veth peer name veth2")
    os.system("sudo ip link del veth3 type veth peer name veth4")

    # Add and "sdon" ovs switch
    os.system("sudo ovs-vsctl add-br sdon")
    os.system("sudo ovs-vsctl add-br mgm")

    # Set eth0 to have no IP address
    os.system("sudo ifconfig eth0 0")

    # Add eth0 to this bridge
    os.system("sudo ovs-vsctl add-port mgm eth0")

    # Get an IP address for sdon using dhclient
    os.system("sudo dhclient mgm -v")

    f = os.popen('ifconfig mgm | grep "inet\ addr" | cut -d: -f2 | cut -d" " -f1')
    global MY_PUB_IP
    MY_PUB_IP = f.read().__str__().strip()
    print "My public IP is:" + MY_PUB_IP

    # Set "sdon" OF version to 1.3 so it can talk to Ryu
    os.system("sudo ovs-vsctl set bridge sdon protocols=OpenFlow13")

    # Create an interface pair for SdonManager
    os.system("sudo ip link add veth1 type veth peer name veth2")

    # Add veth1 to sdon
    os.system("sudo ovs-vsctl add-port sdon veth1")

    # Set veth1 and veth2 up
    os.system("sudo ifconfig veth1 up")
    os.system("sudo ifconfig veth2 up")

    # Set ipv4 forwarding to Enabled
    os.system("sudo sysctl -w net.ipv4.ip_forward=1")

    # Set remote controller
    os.system("sudo ovs-vsctl set-controller sdon tcp:" + CTRL_IP + ":" + CTRL_PORT)
    time.sleep(1)

    # Prepare and send an hello message to the controller
    data = MY_PUB_IP + ",hello"
    sendp(Ether()/IP(dst="1.2.3.4")/data, iface=CTRL_INTF)

    # Wait for the controller to acknowledge the packet (should happen in 3 seconds)
    pkt = sniff(count=1, filter="host 1.2.3.4", timeout=1, iface=CTRL_INTF)
    assert handleSdonSignalingPkt(pkt)


def getMyIp():
    print MY_PUB_IP
    pkt = Ether()/IP(src=str(MY_PUB_IP), dst="1.2.3.4")/UDP()/(MY_PUB_IP + ",getIp")
    print pkt
    sendp(pkt, iface=CTRL_INTF)
    pkt = sniff(count=1, filter="host 1.2.3.4", timeout=1, iface=CTRL_INTF)
    print pkt
    assert handleSdonSignalingPkt(pkt)


def getMyRole():
    sendp(Ether()/IP(src=MY_PUB_IP, dst="1.2.3.4")/UDP()/(MY_PUB_IP + ",getRole"), iface=CTRL_INTF)
    pkt = sniff(count=1, filter="host 1.2.3.4", timeout=1, iface=CTRL_INTF)
    assert handleSdonSignalingPkt(pkt)


def sendTestPacket():
    sendp(Ether()/IP(src=MY_PUB_IP, dst="1.2.3.4")/UDP()/(MY_PUB_IP + "TEST_TEST_TEST_TEST"), iface=CTRL_INTF)


def getMyTunnels():
    sendp(Ether()/IP(src=MY_PUB_IP, dst="1.2.3.4")/UDP()/(MY_PUB_IP + ",getTunnels"), iface=CTRL_INTF)
    pkt = sniff(count=1, filter="host 1.2.3.4", timeout=1, iface=CTRL_INTF)
    assert handleSdonSignalingPkt(pkt)

def getMyOfRules():
    sendp(Ether()/IP(src=MY_PUB_IP, dst="1.2.3.4")/UDP()/(MY_PUB_IP + ",getRules"), iface=CTRL_INTF)
    pkt = sniff(count=1, filter="host 1.2.3.4", timeout=1, iface=CTRL_INTF)
    assert handleSdonSignalingPkt(pkt)

# def waitingForInstructions():
#     pkt = sniff(count=1, filter="host 1.2.3.4", timeout=600, iface=CTRL_INTF)
#     assert handleSdonSignalingPkt(pkt)

def getGo():
    sendp(Ether()/IP(src=MY_PUB_IP, dst="1.2.3.4")/UDP()/(MY_PUB_IP + ",getGo"), iface=CTRL_INTF)
    pkt = sniff(count=1, filter="host 1.2.3.4", timeout=300, iface=CTRL_INTF)
    assert handleSdonSignalingPkt(pkt)


# Actions to be done after the Controller has responded
def setupMyIp(ip):
    # Create an interface pair for our File Transfer app
    os.system("sudo ip link add veth3 type veth peer name veth4")

    # Add veth3 to sdon
    os.system("sudo ovs-vsctl add-port sdon veth3")

    # Set veth3 and veth4 up
    os.system("sudo ifconfig veth3 up")
    os.system("sudo ifconfig veth4 up")
    time.sleep(1)
    print "sudo ifconfig veth4 " + str(ip)
    os.system(str("sudo ifconfig veth4 " + str(ip)))

    global MY_IP
    global MY_ID
    ip,mask = ip.split("/")
    MY_IP = ip
    id = ip.split(".")[-1]
    MY_ID = id
    print "My ID is: ", MY_ID

def setupMyTunnels(tunnels):
    greId = 0
    for neighIp in tunnels:
        print "Setting up gre tunnel for: " + neighIp
        # The convention to name the tunnels is >>> my tep IP = 10.lowerId.higherId.MyID/24 <<<
        # and the tunnel destination is >> gre remote IP = 10.lowerId.hirherID.neighID <<
        # if int(MY_ID) < int(neighID):
        #     a = MY_ID
        #     b = neighID
        # else:
        #     a = neighID
        #     b = MY_ID

        # tepIP = "10." + a + "." + b + "." + MY_ID + "/24"
        # neighIp =  "10." + a + "." + b + "." + neighID

        print "Tunnel settings: neighbour Ip=", neighIp

        # os.system("sudo ovs-vsctl add-port mgm tep" + str(greId) + " -- set interface tep" + str(greId) + " type=internal")
        # os.system("sudo ifconfig tep" + str(greId) + " " + tepIP)
        os.system("sudo ovs-vsctl add-port sdon gre" + str(greId) + " -- set interface gre" + str(greId) + " type=gre options:remote_ip=" + str(neighIp))

        greId += 1

def startMyFtApp():
    print "Starting the FT app!"
    print MY_ROLE
    if MY_ROLE == "S":
        print "I'm a sender and will start sending."

        sentMsg = ""

        if MY_IP == "192.168.101.1":
            recIp = "192.168.101.4"
        if MY_IP == "192.168.101.2":
            recIp = "192.168.101.6"

        for i in range(1,4):
            time.sleep(5)
            # Construct packet
            SRC = MY_IP
            DST = recIp
            seqNo = str(i)
            state = "ORG" # this is an original packet, the adder will change it to ADD
            isLast = "F"
            if i == 3:
                isLast = "T"

            data = "".join( [random.choice(string.letters[:26]) for i in xrange(8)] )
            sentMsg += data
            # data = "DumyData" * (16/8) # 8 bytes * 512/8 = 512 bytes in each packet
            payload = "/".join([SRC,DST,seqNo,state,isLast,data])
            pkt = Ether(dst="aa:1d:15:6f:14:cf")/IP(src=MY_IP, dst=recIp)/payload
            print "Sending packet: "
            print payload
            sendp(pkt, iface=APP_INTF)

        print "I've send: ", sentMsg

    if MY_ROLE == "D":
        # packets = []
        orgMsg = ""
        addMsg = ""
        lastFlagCount = 0
        last = False
        while not last:
            filter = "host 192.168.101.4 or host 192.168.101.6"
            pkt = sniff(count=1, filter=filter, iface=APP_INTF)
            # pkt = sniff(count=1, filter=("host " + MY_IP), iface=APP_INTF)
            print "I'm a Destination and got this packet: "
            pkt = pkt[0] # sniff outputs packets as list
            # print pkt
            pl = str(pkt[IP].payload)
            print "Payload is: ", pl
            # print "Type of pl is: ", type(pl)

            SRC, DST, seqNo, state, isLast, data = pl.split("/")

            if state == "ORG":
                orgMsg += data
            if state == "ADD":
                addMsg += data

            if isLast == "T":
                lastFlagCount += 1
            if lastFlagCount == 2:
                last= True

            # packets.append(pkt)

        print "Original message is:", orgMsg
        print "Coded    message is:", addMsg
        print "     ===XOR===      "
        print "Other    message is:", sxor(orgMsg,addMsg)

    if MY_ROLE == "F":
        # All I have to do is echo packets on the network
        lastFlagCount = 0
        last = False
        while not last:
            filter = "host 192.168.101.4 or host 192.168.101.6"
            pkt = sniff(count=1, filter=filter, iface=APP_INTF)
            print "I'm a forwarder, and am bouncing this packet:"
            pkt = pkt[0] # sniff outputs packets as list
            # print pkt
            payload = str(pkt[IP].payload)
            print payload

            SRC, DST, seqNo, state, isLast, data = payload.split("/")

            sendp(pkt, iface=APP_INTF)

            if isLast == "T":
                lastFlagCount += 1
            if lastFlagCount == 1:
                last= True


    if MY_ROLE == "A":
        # My role is more complicated:
        # I receive packets, get their seqNo, and if I have a matching one, I'll
        packets = {"192.168.101.1": {}, # this is for A
                   "192.168.101.2": {}} # this is for B
        last = False
        lastFlagCount = 0
        while not last:
            filter = "host 192.168.101.4 or host 192.168.101.6"
            pkt = sniff(count=1, filter=filter, iface=APP_INTF)
            print "I'm an adding and bouncing this packet:"
            pkt = pkt[0] # sniff outputs packets as list
            # print pkt
            payload = str(pkt[IP].payload)
            print "Payload is: ", payload
            # print "Type of payload is: ", type(payload)
            SRC, DST, seqNo, state, isLast, data = payload.split("/")

            # add packet details to dictionary
            packets[SRC][seqNo]=[SRC, DST, state, isLast, data]

            # check if this was the last packet of the session
            if isLast == "T":
                lastFlagCount += 1
            if lastFlagCount == 2:
                last= True

            # check if the current seqNo is already in the list of the other sender;
            pairIp = ""
            for i in packets.keys(): # Get the pair's IP address
                if SRC != i:
                    pairIp = i
                    print pairIp

            try: # if there is an entry for the pair of this packet
                pSRC, pDST, pState, pIsLast, pData = packets[pairIp][seqNo]

                # commonData = data + pData # here I have to do it with the actual XOR function
                commonData = sxor(data, pData)

                payload = "/".join([SRC,DST,seqNo,"ADD",isLast,commonData])
                # pairPayload = "/".join([pSRC,pDST,seqNo,"ADD",pIsLast,commonData])

                pkt = Ether(dst="aa:1d:15:6f:14:cf")/IP(src=SRC, dst=DST)/payload
                print pkt
                # pairPkt = Ether(dst="aa:1d:15:6f:14:cf")/IP(src=pSRC, dst=pDST)/pairPayload
                # print pairPkt

                # Now send the packets
                sendp(pkt, iface=APP_INTF)
                # time.sleep(2)
                # sendp(pairPkt, iface=APP_INTF)

            except KeyError:
                print "Yet, there is no pair entry for this seqNo: ", seqNo


def handleControllerReply(msg):
    action, context = msg.split(",")
    if action == "hello":
        print "Controller acknowledged my hello message."

    if action == "setIp":
        print "Controller wants me to start FT at: " + context
        ip = context
        setupMyIp(ip)

    if action == "setRoles":
        print "Controller wants me to start FT as: " + context
        roles = context.strip("[]")
        roles = roles.split(";")
        global MY_ROLE
        MY_ROLE = roles[0]
        print MY_ROLE

    if action == "setTunnels":
        print "Controller wants me create tunnels for: " + context
        tunnels = context.strip("[]")
        tunnels = tunnels.split(";")
        setupMyTunnels(tunnels)

    if action == "setRules":
        print "Controller acknowledged my getRules message."

    if action == "setGo":
        print "Controller confirmed I can start my FT app."
        startMyFtApp()

def sxor(s1,s2):
    # convert strings to a list of character pair tuples
    # go through each tuple, converting them to ASCII code (ord)
    # perform exclusive or on the ASCII code
    # then convert the result back to ASCII (chr)
    # merge the resulting array of characters as a string
    return ''.join(chr(ord(a) ^ ord(b)) for a,b in zip(s1,s2))

mainMenu()
